README.md
ewp-rsa-aes-java Library
========================

Java implementation of the [`ewp-rsa-aes128gcm` encryption][encr-spec].


Installation and Documentation
------------------------------

Requires **Java 8 SE**.

Releases are deployed to *Maven Central Repository*. You'll simply need to
include a proper reference in your build's dependencies. Click the image below
for the artifact details.

[
    ![Maven Central](https://maven-badges.herokuapp.com/maven-central/eu.erasmuswithoutpaper/ewp-rsa-aes/badge.svg)
](https://maven-badges.herokuapp.com/maven-central/eu.erasmuswithoutpaper/ewp-rsa-aes)

You can also browse the project's Javadocs here:

[
    ![Javadocs](http://javadoc.io/badge/eu.erasmuswithoutpaper/ewp-rsa-aes.svg?color=red)
](http://javadoc.io/doc/eu.erasmuswithoutpaper/ewp-rsa-aes)

**Upgrading?** Check out the [changelog (release notes)](CHANGELOG.md).


Versioning strategy
-------------------

We use [semantic versioning](http://semver.org/) (`MAJOR.MINOR.PATCH`) for our
release version numbers.

 * **Major version** is incremented when our changes are likely to break your
   builds or runtime behavior.

 * **Minor version** is incremented when new features are added. (Note, that
   such changes still can break your builds in some rare cases.)

 * **Patch version** is incremented on bug fixes, documentation updates, etc.


[encr-spec]: https://github.com/erasmus-without-paper/ewp-specs-sec-rsa-aes128gcm


CHANGELOG.md
Release notes
=============

1.0.1
-----

* Bugfix: Static `EwpRsaAes128GcmDecoder` methods returned truncated SHA-256
  hashes. Only the first 16 bytes were returned (instead of the full 32 bytes).


1.0.0
-----

*Released on 2017-12-13*

First official release.


src/test/java/eu/erasmuswithoutpaper/rsaaes/TestBase.java
package eu.erasmuswithoutpaper.rsaaes;

import java.io.IOException;
import java.io.InputStream;
import java.io.UnsupportedEncodingException;

import org.apache.commons.io.IOUtils;

/**
 * A common base for all other test classes. Provides some useful shortcut methods.
 */
public class TestBase {

  /**
   * Quick way of fetching files from resources.
   *
   * @param path A path relative to "test-files" directory. The file must exist.
   * @return The contents of the file.
   */
  protected static byte[] getFile(String path) {
    try {
      return getPossiblyNonExistingFile(path);
    } catch (IOException e) {
      throw new RuntimeException(e);
    }
  }

  /**
   * Same as {@link #getFile(String)}, but converts the file to String.
   *
   * @param path as in {@link #getFile(String)}.
   * @return Contents transformed to a string (with UTF-8 encoding).
   */
  protected static String getFileAsString(String path) {
    byte[] bytes = getFile(path);
    try {
      return new String(bytes, "utf-8");
    } catch (UnsupportedEncodingException e) {
      throw new RuntimeException(e);
    }
  }

  /**
   * Same as {@link #getFile(String)}, but the file is not required to exist.
   *
   * @param path as in {@link #getFile(String)}.
   * @throws IOException if the file does not exist.
   */
  protected static byte[] getPossiblyNonExistingFile(String path) throws IOException {
    InputStream stream = TestBase.class.getResourceAsStream("/test-files/" + path);
    if (stream == null) {
      throw new IOException("No such resource");
    }
    return IOUtils.toByteArray(stream);
  }
}


src/test/java/eu/erasmuswithoutpaper/rsaaes/UtilsTest.java
package eu.erasmuswithoutpaper.rsaaes;

import static org.assertj.core.api.Assertions.assertThat;

import org.junit.Test;

import eu.erasmuswithoutpaper.rsaaes.Utils;

public class UtilsTest extends TestBase {

  @Test
  public void testAddLineBreaks() {
    assertThat(Utils.addLineBreaks("abc", 1)).isEqualTo("a\nb\nc");
    assertThat(Utils.addLineBreaks("abc", 2)).isEqualTo("ab\nc");
    assertThat(Utils.addLineBreaks("abc", 3)).isEqualTo("abc");
    assertThat(Utils.addLineBreaks("abc", 4)).isEqualTo("abc");

    assertThat(Utils.addLineBreaks("a", 1)).isEqualTo("a");
    assertThat(Utils.addLineBreaks("a", 2)).isEqualTo("a");

    assertThat(Utils.addLineBreaks("", 1)).isEqualTo("");
    assertThat(Utils.addLineBreaks("", 2)).isEqualTo("");
  }

}


src/test/java/eu/erasmuswithoutpaper/rsaaes/EwpRsaAes128GcmTest.java
package eu.erasmuswithoutpaper.rsaaes;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.core.api.Assertions.fail;

import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.NoSuchAlgorithmException;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.security.spec.InvalidKeySpecException;
import java.security.spec.PKCS8EncodedKeySpec;
import java.security.spec.X509EncodedKeySpec;
import java.util.Arrays;

import org.junit.BeforeClass;
import org.junit.Test;

import eu.erasmuswithoutpaper.rsaaes.EwpRsaAes128GcmDecoder.DecryptionInternals;

public class EwpRsaAes128GcmTest extends TestBase {

  private static RSAPublicKey recipientPublicKey;
  private static RSAPrivateKey recipientPrivateKey;

  @BeforeClass
  public static void setUpClass() {
    try {
      KeyFactory kf = KeyFactory.getInstance("RSA");
      recipientPrivateKey = (RSAPrivateKey) kf.generatePrivate(
          new PKCS8EncodedKeySpec(Utils.b64decode(getFileAsString("rsa1-private"))));
      recipientPublicKey = (RSAPublicKey) kf
          .generatePublic(new X509EncodedKeySpec(Utils.b64decode(getFileAsString("rsa1-public"))));
    } catch (NoSuchAlgorithmException | InvalidKeySpecException e) {
      throw new RuntimeException(e);
    }
  }

  @Test
  public void generateNewExampleForSpecWorks() throws BadEwpRsaAesBody, InvalidRecipient {
    String example = this.generateNewExampleForSpec();
    // assertThat(example).isEqualTo("Something");
  }

  @Test
  public void testBasicUsage() throws BadEwpRsaAesBody, InvalidRecipient {
    EwpRsaAes128GcmEncoder encoder = new EwpRsaAes128GcmEncoder(recipientPublicKey);
    byte[] payload = "This is a secret".getBytes(StandardCharsets.UTF_8);
    byte[] ewpRsaAesBody = encoder.encode(payload);

    EwpRsaAes128GcmDecoder decoder =
        new EwpRsaAes128GcmDecoder(recipientPublicKey, recipientPrivateKey);
    byte[] decryptedPayload = decoder.decode(ewpRsaAesBody);
    assertThat(decryptedPayload).isEqualTo(payload);
  }

  @Test
  public void testErrors() {

    EwpRsaAes128GcmDecoder decoder =
        new EwpRsaAes128GcmDecoder(recipientPublicKey, recipientPrivateKey);

    byte[] ewpRsaAesBody;

    /* Bytes near the end were changed. Should fail during GCM AuthTag checking. */

    ewpRsaAesBody = Utils.b64decode(getFileAsString("example1-ewpRsaAesBody-broken1"));
    try {
      decoder.decode(ewpRsaAesBody);
      fail("Exception expected, but not thrown");
    } catch (BadEwpRsaAesBody e) {
      // Expected.
    } catch (InvalidRecipient e) {
      throw new RuntimeException(e);
    }

    /*
     * Bytes in the RSA-encrypted section are changed. Should fail during RSA decryption, or AES key
     * construction.
     */

    ewpRsaAesBody = Utils.b64decode(getFileAsString("example1-ewpRsaAesBody-broken2"));
    try {
      decoder.decode(ewpRsaAesBody);
      fail("Exception expected, but not thrown");
    } catch (BadEwpRsaAesBody e) {
      // Expected.
    } catch (InvalidRecipient e) {
      throw new RuntimeException(e);
    }

    /* Bytes are appended to an otherwise valid body. Should fail during GCM AuthTag checking. */

    ewpRsaAesBody = Utils.b64decode(getFileAsString("example1-ewpRsaAesBody-appended"));
    try {
      decoder.decode(ewpRsaAesBody);
      fail("Exception expected, but not thrown");
    } catch (BadEwpRsaAesBody e) {
      // Expected.
    } catch (InvalidRecipient e) {
      throw new RuntimeException(e);
    }

    /*
     * Bytes are changed in the beginning, so that the recipient ID should not match the one used by
     * our decoder.
     */

    ewpRsaAesBody = Utils.b64decode(getFileAsString("example1-ewpRsaAesBody-badrecipient"));
    try {
      decoder.decode(ewpRsaAesBody);
      fail("Exception expected, but not thrown");
    } catch (BadEwpRsaAesBody e) {
      throw new RuntimeException(e);
    } catch (InvalidRecipient e) {
      // Expected.
    }

  }

  @Test
  public void testExampleFromSpec() throws BadEwpRsaAesBody, InvalidRecipient {
    byte[] ewpRsaAesBody = Utils.b64decode(getFileAsString("example1-ewpRsaAesBody"));

    EwpRsaAes128GcmDecoder decoder =
        new EwpRsaAes128GcmDecoder(recipientPublicKey, recipientPrivateKey);
    DecryptionInternals details = decoder.decodeWithDetails(ewpRsaAesBody);

    // recipientPublicKeyFingerprint

    assertThat(Utils.b64encode(details.recipientPublicKeySha256))
        .isEqualTo("A1ATd09ZbhiHNEvaigZGIDB1lZI1XbP1HISY/9Cxit0=");

    // Also test the corresponding static method

    assertThat(EwpRsaAes128GcmDecoder.extractRecipientPublicKeySha256(ewpRsaAesBody))
        .isEqualTo(details.recipientPublicKeySha256);
    assertThat(EwpRsaAes128GcmDecoder.extractRecipientPublicKeySha256Base64(ewpRsaAesBody))
        .isEqualTo("A1ATd09ZbhiHNEvaigZGIDB1lZI1XbP1HISY/9Cxit0=");
    assertThat(EwpRsaAes128GcmDecoder.extractRecipientPublicKeySha256Hex(ewpRsaAesBody))
        .isEqualTo("035013774f596e1887344bda8a06462030759592355db3f51c8498ffd0b18add");

    // aesKey

    assertThat(details.encryptedAesKeyLength).isEqualTo(256); // 2048 bits (size of the RSA key)
    assertThat(Utils.b64encode(details.encryptedAesKey))
        .isEqualTo(getFileAsString("example1-encryptedAesKey"));
    assertThat(details.aesKey.length).isEqualTo(16);
    assertThat(Utils.b64encode(details.aesKey)).isEqualTo("Gwaty5wYxuD81f++z6jwZw==");

    // iv

    assertThat(Utils.b64encode(details.iv)).isEqualTo("t8/I35bQAcG6YpXk");

    // encryptedPayload

    assertThat(details.encryptedPayloadOffset).isEqualTo(302);
    byte[] encryptedPayload =
        Arrays.copyOfRange(ewpRsaAesBody, details.encryptedPayloadOffset, ewpRsaAesBody.length);
    assertThat(Utils.b64encode(encryptedPayload))
        .isEqualTo("tIsU5R2OToxjUqI2V/vFyNAuArAkFVK4TLI0Tk5Vo5yN");

    // payload

    assertThat(details.payload.length).isEqualTo(17);
    assertThat(Utils.b64encode(details.payload)).isEqualTo("VGhpcyBpcyBhIHNlY3JldC4=");
    assertThat(new String(details.payload, StandardCharsets.UTF_8)).isEqualTo("This is a secret.");
  }

  private String generateNewExampleForSpec() throws BadEwpRsaAesBody, InvalidRecipient {

    StringBuilder sb = new StringBuilder();
    sb.append("// recipientPublicKey\n");
    sb.append(Utils.b64encode(recipientPublicKey.getEncoded())).append("\n\n");
    sb.append("// recipientPrivateKey\n");
    sb.append(Utils.b64encode(recipientPrivateKey.getEncoded())).append("\n\n");

    byte[] originalContent = "This is a secret.".getBytes(StandardCharsets.UTF_8);
    EwpRsaAes128GcmEncoder encoder = new EwpRsaAes128GcmEncoder(recipientPublicKey);
    byte[] ewpRsaAesBody = encoder.encode(originalContent);
    sb.append("// ewpRsaAesBody\n");
    sb.append(Utils.b64encode(ewpRsaAesBody)).append("\n\n");

    EwpRsaAes128GcmDecoder decoder =
        new EwpRsaAes128GcmDecoder(recipientPublicKey, recipientPrivateKey);
    DecryptionInternals details = decoder.decodeWithDetails(ewpRsaAesBody);
    sb.append("// recipientPublicKeyFingerprint (base64)\n");
    sb.append(Utils.b64encode(details.recipientPublicKeySha256)).append("\n\n");
    sb.append("// encryptedAesKeyLength\n");
    sb.append(details.encryptedAesKeyLength).append("\n\n");
    sb.append("// encryptedAesKey (base64)\n");
    sb.append(Utils.b64encode(details.encryptedAesKey)).append("\n\n");
    sb.append("// aesKey (base64)\n");
    sb.append(Utils.b64encode(details.aesKey)).append("\n\n");
    sb.append("// iv (base64)\n");
    sb.append(Utils.b64encode(details.iv)).append("\n\n");
    sb.append("// encryptedPayloadOffset\n");
    sb.append(details.encryptedPayloadOffset).append("\n\n");
    byte[] encryptedPayload =
        Arrays.copyOfRange(ewpRsaAesBody, details.encryptedPayloadOffset, ewpRsaAesBody.length);
    sb.append("// encryptedPayload (base64)\n");
    sb.append(Utils.b64encode(encryptedPayload)).append("\n\n");
    sb.append("// payload (base64)\n");
    sb.append(Utils.b64encode(details.payload)).append("\n\n");
    sb.append("// payload (UTF-8)\n");
    sb.append(new String(details.payload, StandardCharsets.UTF_8)).append("\n\n");

    String example = sb.toString();
    return example;
  }
}


src/main/java/eu/erasmuswithoutpaper/rsaaes/Utils.java
package eu.erasmuswithoutpaper.rsaaes;

import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.Base64;

class Utils {

  private static final char[] HEX_CHARS = "0123456789abcdef".toCharArray();
  private static final byte[] LF = new byte[] { 0x0A };

  static String addLineBreaks(String str, int lineLength) {
    StringBuilder sb = new StringBuilder();
    int offset = 0;
    while (true) {
      int nextOffset = offset + lineLength;
      if (nextOffset < str.length()) {
        sb.append(str.substring(offset, nextOffset));
        sb.append('\n');
        offset = nextOffset;
      } else {
        sb.append(str.substring(offset));
        break;
      }
    }
    return sb.toString();
  }

  static byte[] b64decode(String encoded) {
    return Base64.getMimeDecoder().decode(encoded);
  }

  static String b64encode(byte[] data) {
    return Base64.getMimeEncoder(76, LF).encodeToString(data);
  }

  static byte[] getBinarySha256Fingerprint(byte[] data) {
    try {
      return MessageDigest.getInstance("SHA-256").digest(data);
    } catch (NoSuchAlgorithmException e) {
      throw new RuntimeException(e);
    }
  }

  static String hexEncode(byte[] data) {
    char[] chars = new char[2 * data.length];
    for (int i = 0; i < data.length; ++i) {
      chars[2 * i] = HEX_CHARS[(data[i] & 0xF0) >>> 4];
      chars[2 * i + 1] = HEX_CHARS[data[i] & 0x0F];
    }
    return new String(chars);
  }
}


src/main/java/eu/erasmuswithoutpaper/rsaaes/EwpRsaAes128GcmEncoder.java
package eu.erasmuswithoutpaper.rsaaes;

import java.nio.ByteBuffer;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.SecureRandom;
import java.security.interfaces.RSAPublicKey;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.KeyGenerator;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;

/**
 * Encoder for the `ewp-rsa-aes128gcm` encryption, as defined here:
 * https://github.com/erasmus-without-paper/ewp-specs-sec-rsa-aes128gcm
 *
 * <p>
 * This class in thread-safe.
 * </p>
 */
public class EwpRsaAes128GcmEncoder {

  protected final RSAPublicKey recipientPublicKey;
  protected final byte[] recipientPublicKeyId;
  protected final SecretKey aesKey;
  protected final byte[] aesKeyEncrypted;

  /**
   * Create a new encoder for a specified recipient. This encoder can be reused for this recipient
   * (and only for this recipient).
   *
   * @param recipientPublicKey the public key of the recipient. It will be used to encrypt the
   *        underlying AES key (which in turn will be used for the actual payload encryption).
   */
  public EwpRsaAes128GcmEncoder(RSAPublicKey recipientPublicKey) {
    this.recipientPublicKey = recipientPublicKey;
    this.recipientPublicKeyId = Utils.getBinarySha256Fingerprint(recipientPublicKey.getEncoded());
    this.aesKey = this.generateNewAesKey();
    this.aesKeyEncrypted = this.encryptAesKey(this.aesKey);
  }

  /**
   * Encrypt the given payload and return the encoded result.
   *
   * @param body Payload to be encrypted.
   * @return Encrypted payload encoded in the <code>ewpRsaAesBody</code> format.
   */
  public byte[] encode(byte[] body) {

    // ByteBuffer buf = this.allocateBuffer(body);

    int bufferSize = 0;

    // SHA-256 hash of the recipient's public key (in binary).

    bufferSize += 32;

    // Integer, the length of the encrypted symmetric key.

    bufferSize += 2;

    // The encrypted AES key.

    bufferSize += this.aesKeyEncrypted.length;

    // AES initialization vector

    byte[] iv = this.generateNewIv();
    bufferSize += this.getGcmIvSize();

    // The encrypted body

    byte[] encryptedBody;
    Cipher cipherAes;
    try {
      cipherAes = Cipher.getInstance("AES/GCM/NoPadding");
    } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
      throw new MissingFeature(e);
    }
    GCMParameterSpec spec = new GCMParameterSpec(this.getGcmTagLength() * 8, iv);
    try {
      cipherAes.init(Cipher.ENCRYPT_MODE, this.aesKey, spec);
    } catch (InvalidKeyException | InvalidAlgorithmParameterException e) {
      throw new RuntimeException(e);
    }
    try {
      encryptedBody = cipherAes.doFinal(body);
    } catch (IllegalBlockSizeException | BadPaddingException e) {
      throw new RuntimeException(e);
    }
    bufferSize += encryptedBody.length;

    // Glue everything together.

    ByteBuffer buf = ByteBuffer.allocate(bufferSize);
    buf.put(this.recipientPublicKeyId);
    buf.putShort((short) this.aesKeyEncrypted.length);
    buf.put(this.aesKeyEncrypted);
    buf.put(iv);
    buf.put(encryptedBody);
    return buf.array();
  }

  /**
   * @return The public key of the recipient for which this encoder instance encrypts payloads for.
   */
  public RSAPublicKey getRecipientPublicKey() {
    return this.recipientPublicKey;
  }

  /**
   * @return SHA-256 fingerprint of the recipient's public key.
   */
  public byte[] getRecipientPublicKeySha256() {
    return this.recipientPublicKeyId.clone();
  }

  /**
   * @return Same as {@link #getRecipientPublicKeySha256()}, but formatted in Base64.
   */
  public String getRecipientPublicKeySha256Base64() {
    return Utils.b64encode(this.recipientPublicKeyId);
  }

  /**
   * @return Same as {@link #getRecipientPublicKeySha256()}, but formatted in HEX.
   */
  public String getRecipientPublicKeySha256Hex() {
    return Utils.hexEncode(this.recipientPublicKeyId);
  }

  @Override
  public String toString() {
    return "EwpRsaAes128GcmEncoder[recipient=" + this.getRecipientPublicKeySha256Hex() + "]";
  }

  protected byte[] encryptAesKey(SecretKey aesKey) {
    Cipher cipher;
    try {
      cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
    } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
      throw new MissingFeature(e);
    }
    try {
      cipher.init(Cipher.ENCRYPT_MODE, this.recipientPublicKey);
    } catch (InvalidKeyException e) {
      throw new RuntimeException(e);
    }
    try {
      return cipher.doFinal(aesKey.getEncoded());
    } catch (IllegalBlockSizeException | BadPaddingException e) {
      throw new RuntimeException(e);
    }
  }

  protected SecretKey generateNewAesKey() {
    try {
      KeyGenerator keyGen = KeyGenerator.getInstance("AES");
      keyGen.init(this.getAesKeySize());
      return keyGen.generateKey();
    } catch (NoSuchAlgorithmException e) {
      throw new MissingFeature(e);
    }
  }

  protected byte[] generateNewIv() {
    SecureRandom rnd = new SecureRandom();
    byte[] iv = new byte[this.getGcmIvSize()];
    rnd.nextBytes(iv);
    return iv;
  }

  protected int getAesKeySize() {
    return 128;
  }

  protected int getGcmIvSize() {
    return 12;
  }

  protected int getGcmTagLength() {
    return 16;
  }
}


src/main/java/eu/erasmuswithoutpaper/rsaaes/EwpRsaAes128GcmDecoder.java
package eu.erasmuswithoutpaper.rsaaes;

import java.nio.BufferUnderflowException;
import java.nio.ByteBuffer;
import java.security.InvalidAlgorithmParameterException;
import java.security.InvalidKeyException;
import java.security.NoSuchAlgorithmException;
import java.security.interfaces.RSAPrivateKey;
import java.security.interfaces.RSAPublicKey;
import java.util.Arrays;
import java.util.Collections;
import java.util.Map;

import javax.crypto.BadPaddingException;
import javax.crypto.Cipher;
import javax.crypto.IllegalBlockSizeException;
import javax.crypto.NoSuchPaddingException;
import javax.crypto.SecretKey;
import javax.crypto.spec.GCMParameterSpec;
import javax.crypto.spec.SecretKeySpec;

/**
 * Decoder for the `ewp-rsa-aes128gcm` encryption, as defined here:
 * https://github.com/erasmus-without-paper/ewp-specs-sec-rsa-aes128gcm
 *
 * <p>
 * This class in thread-safe.
 * </p>
 */
public class EwpRsaAes128GcmDecoder {

  /**
   * A set of both intermediate and final values used during the decryption process. This is usually
   * not needed for the caller, but can be used for debugging and unit-testing.
   */
  static class DecryptionInternals {
    byte[] recipientPublicKeySha256;
    int encryptedAesKeyLength;
    byte[] encryptedAesKey;
    byte[] aesKey;
    SecretKey aesKeyImpl;
    byte[] iv;
    int encryptedPayloadOffset;
    byte[] payload;
  }

  /**
   * Given the encrypted payload encoded in the <code>ewpRsaAesBody</code> format, extract the
   * SHA-256 fingerprint of the payload's recipient's public key. If multiple recipient keys are
   * used, then this method can be used to pick a proper keypair before
   * {@link EwpRsaAes128GcmDecoder} instance is instantiated.
   *
   * @param ewpRsaAesBody The encrypted payload, encoded in <code>ewpRsaAesBody</code> format.
   * @return 16-byte-long binary SHA-256 fingerprint of the payload's recipient's RSA public key.
   * @throws BadEwpRsaAesBody When ewpRsaAesBody doesn't seem to be valid.
   */
  public static byte[] extractRecipientPublicKeySha256(byte[] ewpRsaAesBody)
      throws BadEwpRsaAesBody {
    try {
      return Arrays.copyOf(ewpRsaAesBody, 32);
    } catch (BufferUnderflowException e) {
      throw new BadEwpRsaAesBody(e);
    }
  }

  /**
   * Same as {@link #extractRecipientPublicKeySha256(byte[])}, but the result is Base64-encoded.
   *
   * @param ewpRsaAesBody The encrypted payload, encoded in <code>ewpRsaAesBody</code> format.
   * @return Base64-encoded SHA-256 fingerprint of the payload's recipient's RSA public key.
   */
  public static String extractRecipientPublicKeySha256Base64(byte[] ewpRsaAesBody) {
    return Utils.b64encode(Arrays.copyOf(ewpRsaAesBody, 32));
  }

  /**
   * Same as {@link #extractRecipientPublicKeySha256(byte[])}, but the result is hex-encoded.
   *
   * @param ewpRsaAesBody The encrypted payload, encoded in <code>ewpRsaAesBody</code> format.
   * @return Hex-encoded SHA-256 fingerprint of the payload's recipient's RSA public key.
   */
  public static String extractRecipientPublicKeySha256Hex(byte[] ewpRsaAesBody) {
    return Utils.hexEncode(Arrays.copyOf(ewpRsaAesBody, 32));
  }

  protected final RSAPublicKey recipientPublicKey;
  protected final byte[] recipientPublicKeyId;
  protected final RSAPrivateKey recipientPrivateKey;
  protected final Map<ByteBuffer, SecretKey> aesKeysCache;

  /**
   * Create a new decoder for a specified recipient. This decoder can be reused for this recipient.
   *
   * @param recipientPublicKey The public key of the recipient.
   * @param recipientPrivateKey The private key of the recipient. It must match the public key.
   */
  public EwpRsaAes128GcmDecoder(RSAPublicKey recipientPublicKey,
      RSAPrivateKey recipientPrivateKey) {
    this.recipientPublicKey = recipientPublicKey;
    this.recipientPublicKeyId = Utils.getBinarySha256Fingerprint(recipientPublicKey.getEncoded());
    this.recipientPrivateKey = recipientPrivateKey;
    this.aesKeysCache = Collections
        .synchronizedMap(new LruCache<ByteBuffer, SecretKey>(this.getAesKeysCacheSize()));
  }

  /**
   * @param ewpRsaAesBody The encrypted payload, encoded in <code>ewpRsaAesBody</code> format.
   * @return True, if this decoder's recipient matches the desired recipient of the given encrypted
   *         payload.
   * @throws BadEwpRsaAesBody When ewpRsaAesBody doesn't seem to be valid.
   */
  public boolean canDecode(byte[] ewpRsaAesBody) throws BadEwpRsaAesBody {
    return Arrays.equals(this.recipientPublicKeyId, extractRecipientPublicKeySha256(ewpRsaAesBody));
  }

  /**
   * Decode the given <code>ewpRsaAesBody</code> content, decrypt it, and return the decrypted
   * payload.
   *
   * @param ewpRsaAesBody The encrypted payload, encoded in <code>ewpRsaAesBody</code> format.
   * @return The decrypted payload.
   * @throws BadEwpRsaAesBody When ewpRsaAesBody doesn't seem to be valid.
   * @throws InvalidRecipient When the payload's recipient doesn't match the recipient for which
   *         this decoder has been created for.
   */
  public byte[] decode(byte[] ewpRsaAesBody) throws BadEwpRsaAesBody, InvalidRecipient {
    return this.decodeWithDetails(ewpRsaAesBody).payload;
  }

  /**
   * @return The public key of the recipient for which this decoder instance decrypts payloads for.
   */
  public RSAPublicKey getRecipientPublicKey() {
    return this.recipientPublicKey;
  }

  /**
   * @return SHA-256 fingerprint of the recipient's public key (for which this decoder instance
   *         decrypts payloads for).
   */
  public byte[] getRecipientPublicKeySha256() {
    return this.recipientPublicKeyId.clone();
  }

  /**
   * @return Same as {@link #getRecipientPublicKeySha256()}, but formatted in Base64.
   */
  public String getRecipientPublicKeySha256Base64() {
    return Utils.b64encode(this.recipientPublicKeyId);
  }

  /**
   * @return Same as {@link #getRecipientPublicKeySha256()}, but formatted in HEX.
   */
  public String getRecipientPublicKeySha256Hex() {
    return Utils.hexEncode(this.recipientPublicKeyId);
  }

  @Override
  public String toString() {
    return "EwpRsaAes128GcmDecoder[recipient=" + this.getRecipientPublicKeySha256Hex() + "]";
  }

  protected int getAesKeysCacheSize() {
    return 20;
  }

  protected int getGcmIvSize() {
    return 12;
  }

  protected int getGcmTagLength() {
    return 16;
  }

  DecryptionInternals decodeWithDetails(byte[] ewpRsaAesBody)
      throws BadEwpRsaAesBody, InvalidRecipient {

    DecryptionInternals vars = new DecryptionInternals();
    ByteBuffer buf = ByteBuffer.wrap(ewpRsaAesBody);

    try {

      // Check if the recipient matches.

      vars.recipientPublicKeySha256 = new byte[32];
      buf.get(vars.recipientPublicKeySha256);
      if (!Arrays.equals(this.recipientPublicKeyId, vars.recipientPublicKeySha256)) {
        throw new InvalidRecipient();
      }

      // Read the encrypted symmetric key.

      vars.encryptedAesKeyLength = buf.getShort() & 0xffff;
      vars.encryptedAesKey = new byte[vars.encryptedAesKeyLength];
      buf.get(vars.encryptedAesKey);

      // Check if we have this encryptedAesKey in our cache.

      ByteBuffer cacheKey = ByteBuffer.wrap(vars.encryptedAesKey);
      if (this.aesKeysCache.containsKey(cacheKey)) {
        vars.aesKeyImpl = this.aesKeysCache.get(cacheKey);
        vars.aesKey = vars.aesKeyImpl.getEncoded();
      } else {

        // Not found in our cache. Decrypt the key.

        Cipher cipherRsa;
        try {
          cipherRsa = Cipher.getInstance("RSA/ECB/PKCS1Padding");
        } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
          throw new MissingFeature(e);
        }
        try {
          cipherRsa.init(Cipher.DECRYPT_MODE, this.recipientPrivateKey);
        } catch (InvalidKeyException e) {
          throw new RuntimeException(e);
        }
        try {
          vars.aesKey = cipherRsa.doFinal(vars.encryptedAesKey);
        } catch (IllegalBlockSizeException | BadPaddingException e) {
          throw new BadEwpRsaAesBody(e);
        }
        vars.aesKeyImpl = new SecretKeySpec(vars.aesKey, 0, vars.aesKey.length, "AES");

        // Cache it.

        this.aesKeysCache.put(cacheKey, vars.aesKeyImpl);
      }

      // Read AES initialization vector.

      vars.iv = new byte[this.getGcmIvSize()];
      buf.get(vars.iv);

      // Decrypt the body.

      Cipher cipherAes;
      try {
        cipherAes = Cipher.getInstance("AES/GCM/NoPadding");
      } catch (NoSuchAlgorithmException | NoSuchPaddingException e) {
        throw new MissingFeature(e);
      }
      GCMParameterSpec spec = new GCMParameterSpec(this.getGcmTagLength() * 8, vars.iv);
      try {
        cipherAes.init(Cipher.DECRYPT_MODE, vars.aesKeyImpl, spec);
      } catch (InvalidAlgorithmParameterException e) {
        throw new RuntimeException(e);
      } catch (InvalidKeyException e) {
        throw new BadEwpRsaAesBody(e);
      }
      vars.encryptedPayloadOffset = buf.position();
      try {
        vars.payload = cipherAes.doFinal(ewpRsaAesBody, vars.encryptedPayloadOffset,
            ewpRsaAesBody.length - vars.encryptedPayloadOffset);
      } catch (IllegalBlockSizeException | BadPaddingException e) {
        throw new BadEwpRsaAesBody(e);
      }

      return vars;

    } catch (BufferUnderflowException e) {
      throw new BadEwpRsaAesBody(e);
    }
  }
}
